drivers/mipi_dbi/mipi_dbi_spi.c

```diff
@@ -94,34 +98,107 @@ static int mipi_dbi_spi_write_helper(const struct device *dev,
-			gpio_pin_set_dt(&config->cmd_data, 0);
-			ret = spi_write(config->spi_dev, &dbi_config->config,
-					&buf_set);
-			if (ret < 0) {
-				goto out;
+		if (config->xfr_min_bits == MIPI_DBI_SPI_XFR_8BIT) {
+			/* 4 wire mode is much simpler. We just toggle the
+			 * command/data GPIO to indicate if we are sending
+			 * a command or data (the default).
+			 */
+			buffer.buf = &cmd;
+			buffer.len = sizeof(cmd);
+
+			if (cmd_present) {
+				/* Set CD pin low for command */
+				gpio_pin_set_dt(&config->cmd_data, 0);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
 			}
-		}
 
-		if (len > 0) {
-			buffer.buf = (void *)data_buf;
-			buffer.len = len;
-
-			/* Set CD pin high for data */
-			gpio_pin_set_dt(&config->cmd_data, 1);
-			ret = spi_write(config->spi_dev, &dbi_config->config,
-					&buf_set);
-			if (ret < 0) {
-				goto out;
+			if (len > 0) {
+				buffer.buf = (void *)data_buf;
+				buffer.len = len;
+
+				/* Set CD pin high for data */
+				gpio_pin_set_dt(&config->cmd_data, 1);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
+			}
+		} else if (config->xfr_min_bits == MIPI_DBI_SPI_XFR_16BIT) {
+			/* 4 wire mode with toggle the command/data GPIO
+			 * to indicate if we are sending a command or data
+			 * but send 16-bit blocks (bit stuffing).
+			 */
+			uint16_t data16;
+
+			if (cmd_present) {
+				data16 = sys_cpu_to_be16(cmd);
+				buffer.buf = &data16;
+				buffer.len = sizeof(data16);
+
+				/* Set CD pin low for command */
+				gpio_pin_set_dt(&config->cmd_data, 0);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
+
+				/* Set CD pin high for data, if there are any */
+				if (len > 0) {
+					gpio_pin_set_dt(&config->cmd_data, 1);
+				}
+
+				/* iterate command data */
+				for (int i = 0; i < len; i++) {
+					data16 = sys_cpu_to_be16(data_buf[i]);
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
+			} else {
+				int stuffing = len % sizeof(data16);
+
+				/* Set CD pin high for data, if there are any */
+				if (len > 0) {
+					gpio_pin_set_dt(&config->cmd_data, 1);
+				}
+
+				/* pass through generic device data */
+				if (len - stuffing > 0) {
+					buffer.buf = (void *)data_buf;
+					buffer.len = len - stuffing;
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
+
+				/* iterate remaining data with stuffing */
+				for (int i = len - stuffing; i < len; i++) {
+					data16 = sys_cpu_to_be16(data_buf[i]);
+					buffer.buf = &data16;
+					buffer.len = sizeof(data16);
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
```

Comment:
This block should be moved to separate spi_write_helper_4wire(), or even more finely subdivided.

Reply:
Makes sense, but then also for all other modes - currently 3wire and 4wire - otherwise you would give other developers no clear indication of the intention for breaking down when one mode is handled directly and another mode only with the help of an auxiliary function.

@danieldegrasse: What do you think, should I attach a separate commit with the two additional helper functions `spi_write_helper_3wire()` and `spi_write_helper_4wire()` to this PR, or just refactor the current commit and only provide the single helper function `spi_write_helper_4wire()` for the case of mode `MIPI_DBI_MODE_SPI_4WIRE`?

Reply:
If you are willing to split the code into a helper function `spi_write_helper_3wire` and `spi_write_helper4wire` in this PR, I would prefer that (a separate commit doing this refactoring prior to adding the 16 bit support would be best).

Given the approaching closure of the merge window though, I am ok if only the `spi_write_helper4wire` is done now, and the other change is made as a cleanup PR

Reply:
@danieldegrasse: With the new rebase, there are now two additional commits that split the SPI write function as well as the SPI read functions. In order to maintain consistency within the source, I think it makes sense to split the optional read as well. If this is too much for this PR, I can remove the one commit (read function) and submit it as a separate cleanup PR.

The forked helper functions now only contain the necessary SPI transfer algorithms. The decision as to which help function is necessary and that the device is locked and unlocked is still the responsibility of the SPI read/write function. The new help functions are specified as 'inline' (single point of use) and should potentially avoid a call stack overload.

However, two questions have now arisen with regard to the SPI read functions:

1. the previous implementation always called a `spi_release()` at the end, even if it is never necessary for `MIPI_DBI_MODE_SPI_3WIRE` (missing `SPI_LOCK_ON` or `SPI_HOLD_ON_CS` as with 4wire). This has been retained for the time being. But if it really was an unintentional error in the old code, then I will revise the one commit again.
2. in contrast to the `mipi_dbi_spi_write_helper()`, the `mipi_dbi_spi_command_read()` for the 3wire mode never checked whether an SPI 9-bit transfer is supported by the SPI controller at all. This check is now also included in the `mipi_dbi_spi_read_helper_3wire()`. Would that be okay?